app/src/main/java/org/thoughtcrime/securesms/home/HomeViewModel.kt

```diff
@@ -56,32 +75,55 @@ class HomeViewModel @Inject constructor(
                     .onStart { emit(emptySet()) }
                     .distinctUntilChanged()
 
+    private fun messageRequests() = combine(
+        unapprovedConversationCount(),
+        hasHiddenMessageRequests(),
+        latestUnapprovedConversationTimestamp(),
+        ::createMessageRequests
+    )
+
+    private fun unapprovedConversationCount() = reloadTriggersAndContentChanges()
+        .map { threadDb.unapprovedConversationCount }
+
+    private fun latestUnapprovedConversationTimestamp() = reloadTriggersAndContentChanges()
+        .map { threadDb.latestUnapprovedConversationTimestamp }
+
     @Suppress("OPT_IN_USAGE")
-    private fun observeConversationList(): Flow<List<ThreadRecord>> = merge(
-            manualReloadTrigger,
-            contentResolver.observeChanges(DatabaseContentProviders.ConversationList.CONTENT_URI))
-            .debounce(CHANGE_NOTIFICATION_DEBOUNCE_MILLS)
-            .onStart { emit(Unit) }
-            .mapLatest { _ ->
-                withContext(Dispatchers.IO) {
-                    threadDb.approvedConversationList.use { openCursor ->
-                        val reader = threadDb.readerFor(openCursor)
-                        buildList(reader.count) {
-                            while (true) {
-                                add(reader.next ?: break)
-                            }
-                        }
-                    }
-                }
+    private fun observeConversationList(): Flow<List<ThreadRecord>> = reloadTriggersAndContentChanges()
+        .mapLatest { _ ->
+            threadDb.approvedConversationList.use { openCursor ->
+                threadDb.readerFor(openCursor).run { generateSequence { next }.toList() }
             }
+        }
+
+    @OptIn(FlowPreview::class)
+    private fun reloadTriggersAndContentChanges() = merge(
+        manualReloadTrigger,
+        contentResolver.observeChanges(DatabaseContentProviders.ConversationList.CONTENT_URI)
+    )
+        .flowOn(Dispatchers.IO)
+        .debounce(CHANGE_NOTIFICATION_DEBOUNCE_MILLS)
+        .onStart { emit(Unit) }
 
     fun tryReload() = manualReloadTrigger.tryEmit(Unit)
 
     data class Data(
-            val threads: List<ThreadRecord>,
-            val typingThreadIDs: Set<Long>
+        val threads: List<ThreadRecord> = emptyList(),
+        val typingThreadIDs: Set<Long> = emptySet(),
+        val messageRequests: MessageRequests? = null
     )
 
+    fun createMessageRequests(
+        count: Int,
+        hidden: Boolean,
+        timestamp: Long
+    ) = if (count > 0 && !hidden) MessageRequests(
+        count.toString(),
+        DateUtils.getDisplayFormattedTimeSpanString(context, Locale.getDefault(), timestamp)
+    ) else null
+
+    data class MessageRequests(val count: String, val timestamp: String)
```

Comment:
I don't know how much appetite we have in terms of refactoring this class, I was having this in mind:
Because the message request is also part of the `RecyclerView`, maybe we can have a representative model for the `ViewModel`, like this: 

```kotlin
sealed interface Item {
  data class MessageRequest(val req: MessageRequest) : Item
  data class Thread(val record: ThreadRecord) : Item
}
```


Then we'll have our data as: 
```kotlin
data class Data(
    val items: List<Item>,
    ...
)
```

So that the manual adapter manipulation in the view layer can be skipped!

But this is a bit of long shot so I'm happy if you think we should do that later, separately.

Reply:
Would be reasonable if we kept this in RecyclerView, but in Compose it's probably wouldn't help:

```
ScrollView {
    datamessageRequest?.let { Header(it) }
    LazyThreads(data.threads)
}
```

so perhaps we leave it if it helps for that.